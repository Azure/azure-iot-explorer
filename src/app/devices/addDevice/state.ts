/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { Record } from 'immutable';
import { IM } from '../../shared/types/types';
import { SynchronizationStatus } from '../../api/models/synchronizationStatus';

export interface AddDeviceStateInterface {
    synchronizationStatus: SynchronizationStatus;
}

export const addDeviceStateInitial = Record<AddDeviceStateInterface>({
    synchronizationStatus: SynchronizationStatus.initialized
});

export type AddDeviceStateType = IM<AddDeviceStateInterface>;
