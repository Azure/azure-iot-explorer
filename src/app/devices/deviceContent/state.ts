/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { Record } from 'immutable';
import { IM } from '../../shared/types/types';
import { ModelDefinitionWithSourceWrapper } from '../../api/models/modelDefinitionWithSourceWrapper';
import { DeviceIdentityWrapper } from '../../api/models/deviceIdentityWrapper';
import { DeviceTwinWrapper } from './../../api/models/deviceTwinWrapper';
import { DigitalTwinInterfacePropertiesWrapper } from '../../api/models/digitalTwinInterfacePropertiesWrapper';

export interface DeviceContentStateInterface {
    deviceIdentity: DeviceIdentityWrapper;
    deviceTwin: DeviceTwinWrapper;
    digitalTwinInterfaceProperties: DigitalTwinInterfacePropertiesWrapper;
    interfaceIdSelected: string;
    modelDefinitionWithSource: ModelDefinitionWithSourceWrapper;
}

export const deviceContentStateInitial = Record<DeviceContentStateInterface>({
    deviceIdentity: null,
    deviceTwin: null,
    digitalTwinInterfaceProperties: null,
    interfaceIdSelected: '',
    modelDefinitionWithSource: null,
});

export type DeviceContentStateType = IM<DeviceContentStateInterface>;
