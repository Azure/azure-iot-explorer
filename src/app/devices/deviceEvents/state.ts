/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { Record } from 'immutable';
import { IM } from '../../shared/types/types';
import { SynchronizationWrapper } from '../../api/models/synchronizationWrapper';
import { Message } from '../../api/models/messages';
import { SynchronizationStatus } from '../../api/models/synchronizationStatus';

export interface DeviceEventsStateInterface extends SynchronizationWrapper<Message[]>{}

export const deviceEventsStateInitial = Record<DeviceEventsStateInterface>({
    payload: [],
    synchronizationStatus: SynchronizationStatus.initialized
});

export type DeviceEventsStateType = IM<DeviceEventsStateInterface>;
