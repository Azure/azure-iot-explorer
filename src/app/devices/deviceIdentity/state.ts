/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { Record } from 'immutable';
import { IM } from '../../shared/types/types';
import { DeviceIdentity } from '../../api/models/deviceIdentity';
import { SynchronizationWrapper } from '../../api/models/synchronizationWrapper';
import { SynchronizationStatus } from '../../api/models/synchronizationStatus';

export interface DeviceIdentityStateInterface extends SynchronizationWrapper<DeviceIdentity>{}

export const deviceIdentityStateInitial = Record<DeviceIdentityStateInterface>({
    payload: null,
    synchronizationStatus: SynchronizationStatus.initialized
});

export type DeviceIdentityStateType = IM<DeviceIdentityStateInterface>;
