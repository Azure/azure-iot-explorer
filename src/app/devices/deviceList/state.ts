/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { Record } from 'immutable';
import { IM } from '../../shared/types/types';
import { DeviceSummary } from './../../api/models/deviceSummary';
import { DeviceQuery } from '../../api/models/deviceQuery';
import { SynchronizationStatus } from '../../api/models/synchronizationStatus';

export interface DeviceListStateInterface {
    deviceQuery: DeviceQuery;
    devices: DeviceSummary[];
    synchronizationStatus: SynchronizationStatus;
}

export const deviceListStateInitial = Record<DeviceListStateInterface>({
    deviceQuery: {
        clauses: [],
        continuationTokens: [],
        currentPageIndex: 0,
        deviceId: '',
    },
    devices: [],
    synchronizationStatus: SynchronizationStatus.initialized,
});

export type DeviceListStateType = IM<DeviceListStateInterface>;
