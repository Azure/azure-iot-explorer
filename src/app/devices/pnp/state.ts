/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { Record } from 'immutable';
import { IM } from '../../shared/types/types';
import { ModelDefinitionWithSource } from '../../api/models/modelDefinitionWithSource';
import { SynchronizationWrapper } from '../../api/models/synchronizationWrapper';
import { SynchronizationStatus } from '../../api/models/synchronizationStatus';
import { Twin } from '../../api/models/device';
import { ModuleTwin } from '../../api/models/moduleTwin';

export interface PnpStateInterface {
    twin: SynchronizationWrapper<Twin | ModuleTwin>;
    modelDefinitionWithSource: SynchronizationWrapper<ModelDefinitionWithSource>;
}

export const pnpStateInitial = Record<PnpStateInterface>({
    modelDefinitionWithSource: {
        payload: null,
        synchronizationStatus: SynchronizationStatus.initialized
    },
    twin: {
        payload: null,
        synchronizationStatus: SynchronizationStatus.initialized
    }
});

export type PnpStateType = IM<PnpStateInterface>;
