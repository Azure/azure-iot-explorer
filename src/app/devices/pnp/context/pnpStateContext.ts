/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { pnpStateInitial, PnpStateInterface } from '../state';

// tslint:disable-next-line: no-any
const PnpStateContext = React.createContext<{pnpState: PnpStateInterface, dispatch?: (action: any) => void, getModelDefinition?: () => void}>({ pnpState: pnpStateInitial()});
export const PnpStateContextProvider = PnpStateContext.Provider;
export const usePnpStateContext = () => React.useContext(PnpStateContext);
