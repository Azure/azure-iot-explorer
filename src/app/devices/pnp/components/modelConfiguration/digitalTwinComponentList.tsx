/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { NavLink, useLocation, useRouteMatch } from 'react-router-dom';
import { Announced, IColumn, Label, Pivot, PivotItem, SelectionMode } from '@fluentui/react';
import { ResizableDetailsList } from '../../../../shared/resizeDetailsList/resizableDetailsList';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { usePnpStateContext } from '../../context/pnpStateContext';
import { getComponentNameAndInterfaceIdArray } from '../../utils';
import { getDeviceIdFromQueryString, getModuleIdentityIdFromQueryString } from '../../../../shared/utils/queryStringHelper';
import { ROUTE_PARAMS, ROUTE_PARTS } from '../../../../constants/routes';
import { DEFAULT_COMPONENT_FOR_DIGITAL_TWIN } from '../../../../constants/devices';
import { ErrorBoundary } from '../../../shared/components/errorBoundary';
import { JSONEditor } from '../../../../shared/components/jsonEditor';
import './digitalTwinDetail.scss';

interface ModelContent {
    link: string;
    componentName: string;
    modelId: string;
}

export const DigitalTwinComponentList: React.FC = () => {
    const { t } = useTranslation();
    const { url } = useRouteMatch();
    const { search } = useLocation();
    const { pnpState } = usePnpStateContext();
    const deviceId = getDeviceIdFromQueryString(search);
    const modelDefinitionWithSource = pnpState.modelDefinitionWithSource.payload;
    const modelDefinition = modelDefinitionWithSource && modelDefinitionWithSource.modelDefinition;
    const twin = pnpState.twin.payload;
    const modelId = twin?.modelId;
    const moduleId = getModuleIdentityIdFromQueryString(search);
    const componentNameToIds = getComponentNameAndInterfaceIdArray(modelDefinition);

    const modelContents: ModelContent[]  = componentNameToIds && componentNameToIds.map(nameToId => {
        let link = `${url}${ROUTE_PARTS.DIGITAL_TWINS_DETAIL}/${ROUTE_PARTS.INTERFACES}/` +
            `?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}` +
            `&${ROUTE_PARAMS.COMPONENT_NAME}=${nameToId.componentName}` +
            `&${ROUTE_PARAMS.INTERFACE_ID}=${nameToId.modelId}`;
        if (moduleId) {
            link += `&${ROUTE_PARAMS.MODULE_ID}=${moduleId}`;
        }

        if (nameToId.componentName === DEFAULT_COMPONENT_FOR_DIGITAL_TWIN && nameToId.modelId === modelId) {
            return{
                componentName: t(ResourceKeys.digitalTwin.pivot.defaultComponent),
                link,
                modelId: nameToId.modelId
            };
        }
        else {
            return {
                ...nameToId,
                link
            };
        }
    });

    const getColumns = (): IColumn[] => {
        return [
            {
                fieldName: 'componentName',
                isMultiline: true,
                key: 'name',
                minWidth: 100,
                name: t(ResourceKeys.digitalTwin.componentName)
            },
            {
                fieldName: 'interfaceId',
                isMultiline: true,
                key: 'id',
                minWidth: 100,
                name: t(ResourceKeys.digitalTwin.interfaceId)
            }
        ];
    };

    const renderItemColumn = (item: ModelContent, index: number, column: IColumn) => {
        switch (column.key) {
            case 'name':
                return (
                    <NavLink key={column.key} to={item.link}>
                        {item.componentName}
                    </NavLink>
                );
            case 'id':
                return (
                    <Label
                        key={column.key}
                    >
                        {item.modelId}
                    </Label>
                );
            default:
                return <></>;
        }
    };

    const listView = (
        <>
            {
                modelContents.length !== 0 ?
                    <div className="list-detail">
                        <ResizableDetailsList
                            onRenderItemColumn={renderItemColumn}
                            className="component-list"
                            items={modelContents}
                            columns={getColumns()}
                            selectionMode={SelectionMode.none}
                        />
                    </div> :
                    <>
                        <Label className="no-component">{t(ResourceKeys.digitalTwin.modelContainsNoComponents, {modelId })}</Label>
                        <Announced
                            message={t(ResourceKeys.digitalTwin.modelContainsNoComponents, { modelId })}
                        />
                    </>
            }
        </>);

    return (
        <>
            <h4>{t(ResourceKeys.digitalTwin.steps.third)}</h4>
            <h5>{t(ResourceKeys.digitalTwin.steps.explanation, {modelId})}</h5>
            <Pivot aria-label={t(ResourceKeys.digitalTwin.pivot.ariaLabel)}>
                <PivotItem headerText={t(ResourceKeys.digitalTwin.pivot.components)}>
                    <ErrorBoundary error={t(ResourceKeys.deviceInterfaces.interfaceListFailedToRender)}>
                        {listView}
                    </ErrorBoundary>
                </PivotItem>
                <PivotItem headerText={t(ResourceKeys.digitalTwin.pivot.content)} className="modelContent">
                    <JSONEditor
                        className="interface-definition-json-editor"
                        content={JSON.stringify(modelDefinitionWithSource.modelDefinition, null, '\t')}
                    />
                </PivotItem>
            </Pivot>
        </>
    );
};
