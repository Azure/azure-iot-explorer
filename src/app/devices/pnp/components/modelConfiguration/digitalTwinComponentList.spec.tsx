/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { Announced, Pivot, PivotItem } from '@fluentui/react';
import { DigitalTwinComponentList } from './digitalTwinComponentList';
import { REPOSITORY_LOCATION_TYPE } from '../../../../constants/repositoryLocationTypes';
import { pnpStateInitial, PnpStateInterface } from '../../state';
import * as pnpStateContext from '../../context/pnpStateContext';
import { SynchronizationStatus } from '../../../../api/models/synchronizationStatus';

const interfaceId = 'urn:azureiot:samplemodel;1';

/* tslint:disable */
const deviceTwin: any = {
    "deviceId": "testDevice",
    "modelId": interfaceId,
    "properties" : {
        desired: {
            environmentalSensor: {
                brightness: 456,
                __t: 'c'
            }
        },
        reported: {
            environmentalSensor: {
                brightness: {
                    value: 123,
                    dv: 2
                },
                __t: 'c'
            }
        }
    }
};
/* tslint:enable */

const pathname = 'resources/TestHub.azure-devices.net/devices/deviceDetail/ioTPlugAndPlay/?deviceId=testDevice';

jest.mock('react-router-dom', () => ({
    useHistory: () => ({ push: jest.fn()}),
    useLocation: () => ({ search: '?deviceId=testDevice' }),
    useRouteMatch: () => ({ url: pathname })
}));

describe('DigitalTwinComponentList', () => {
    it('matches snapshot when empty model id is retrieved', () => {
        const initialState: PnpStateInterface = pnpStateInitial().merge({
            twin: {
                payload: {
                    deviceId: 'testDevice',
                    moduleId: ''
                } as any,
                synchronizationStatus: SynchronizationStatus.fetched
            },
            modelDefinitionWithSource: {
                payload: {
                    isModelValid: true,
                    modelDefinition: {
                        '@context': 'dtmi:dtdl:context;2',
                        '@id': 'dtmi:plugnplay:hube2e:cm;1',
                        '@type': 'Interface',
                        'contents': [],
                        'displayName': 'IoT Hub E2E Tests',
                    },
                    source: REPOSITORY_LOCATION_TYPE.Public
                },
                synchronizationStatus: SynchronizationStatus.fetched
            }
        });

        jest.spyOn(pnpStateContext, 'usePnpStateContext').mockReturnValue({ pnpState: initialState, dispatch: jest.fn()});

        const wrapper = shallow(<DigitalTwinComponentList/>);
        expect(wrapper).toMatchSnapshot();
    });

    it('shows model id with valid model definition found but has no component', () => {
        const initialState: PnpStateInterface = pnpStateInitial().merge({
            twin: {
                payload: deviceTwin,
                synchronizationStatus: SynchronizationStatus.fetched
            },
            modelDefinitionWithSource: {
                payload: {
                    isModelValid: true,
                    modelDefinition: {
                        '@context': 'dtmi:dtdl:context;2',
                        '@id': 'dtmi:plugnplay:hube2e:cm;1',
                        '@type': 'Interface',
                        'contents': [],
                        'displayName': 'IoT Hub E2E Tests',
                    },
                    source: REPOSITORY_LOCATION_TYPE.Public
                },
                synchronizationStatus: SynchronizationStatus.fetched
            }
        });

        jest.spyOn(pnpStateContext, 'usePnpStateContext').mockReturnValue({ pnpState: initialState, dispatch: jest.fn()});
        const wrapper = mount(<DigitalTwinComponentList/>);

        expect(wrapper.find(Announced)).toHaveLength(1);
        expect(wrapper.find(Pivot)).toHaveLength(1);
    });

    it('shows model id with valid model definition found and has components', () => {
        const initialState: PnpStateInterface = pnpStateInitial().merge({
            twin: {
                payload: deviceTwin,
                synchronizationStatus: SynchronizationStatus.fetched
            },
            modelDefinitionWithSource: {
                payload: {
                    isModelValid: true,
                    modelDefinition: {
                        '@context': 'dtmi:dtdl:context;2',
                        '@id': 'dtmi:plugnplay:hube2e:cm;1',
                        '@type': 'Interface',
                        'contents': [
                            {
                                '@type': 'Component',
                                'name': 'deviceInformation',
                                'schema': 'dtmi:__DeviceManagement:DeviceInformation;1'
                            },
                            {
                                '@type': 'Component',
                                'name': 'sdkInfo',
                                'schema': 'dtmi:__Client:SDKInformation;1'
                            },
                            {
                                '@type': 'Component',
                                'name': 'environmentalSensor',
                                'schema': 'dtmi:__Contoso:EnvironmentalSensor;1'
                            }
                        ],
                        'displayName': 'IoT Hub E2E Tests',
                    },
                    source: REPOSITORY_LOCATION_TYPE.Public
                },
                synchronizationStatus: SynchronizationStatus.fetched
            }
        });

        jest.spyOn(pnpStateContext, 'usePnpStateContext').mockReturnValue({ pnpState: initialState, dispatch: jest.fn()});
        const wrapper = shallow(<DigitalTwinComponentList/>);

        expect(wrapper.find(Announced)).toHaveLength(0);

        const list = wrapper.find('.component-list');
        expect((list.props() as any).items[0].modelId).toEqual('dtmi:__DeviceManagement:DeviceInformation;1'); // tslint:disable-line:no-any
        expect((list.props() as any).items[1].modelId).toEqual('dtmi:__Client:SDKInformation;1'); // tslint:disable-line:no-magic-numbers, no-any
        expect((list.props() as any).items[2].modelId).toEqual('dtmi:__Contoso:EnvironmentalSensor;1'); // tslint:disable-line:no-magic-numbers, no-any

        // tslint:disable-next-line: no-magic-numbers
        expect(wrapper.find(PivotItem)).toHaveLength(2);
    });
});
