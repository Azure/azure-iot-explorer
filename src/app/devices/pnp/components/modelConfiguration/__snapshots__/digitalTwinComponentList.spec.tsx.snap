// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DigitalTwinComponentList matches snapshot when empty model id is retrieved 1`] = `
<Fragment>
  <h4>
    digitalTwin.steps.third
  </h4>
  <h5>
    digitalTwin.steps.explanation
  </h5>
  <StyledPivot
    aria-label="digitalTwin.pivot.ariaLabel"
  >
    <PivotItem
      headerText="digitalTwin.pivot.components"
    >
      <ErrorBoundary
        error="deviceInterfaces.interfaceListFailedToRender"
      >
        <StyledLabelBase
          className="no-component"
        >
          digitalTwin.modelContainsNoComponents
        </StyledLabelBase>
        <StyledAnnouncedBase
          message="digitalTwin.modelContainsNoComponents"
        />
      </ErrorBoundary>
    </PivotItem>
    <PivotItem
      className="modelContent"
      headerText="digitalTwin.pivot.content"
    >
      <MonacoEditorComponent
        ariaLabel="digitalTwin.pivot.content"
        content="{
	\\"@context\\": \\"dtmi:dtdl:context;2\\",
	\\"@id\\": \\"dtmi:plugnplay:hube2e:cm;1\\",
	\\"@type\\": \\"Interface\\",
	\\"contents\\": [],
	\\"displayName\\": \\"IoT Hub E2E Tests\\"
}"
        height={400}
      />
    </PivotItem>
  </StyledPivot>
</Fragment>
`;
